fmpz/mul_2exp.c

```diff
@@ -9,38 +10,85 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= SMALL_FMPZ_BITCOUNT_MAX)  /* Result fits inside a small fmpz */
+    {
+        if (COEFF_IS_MPZ(*f))
+            _fmpz_clear_mpz(*f);
+
+        *f = c1 << exp;
+    }
+    else if (c1bits <= SMALL_FMPZ_BITCOUNT_MAX)   /* g is small */
+    {
+        ulong expred = exp % GMP_LIMB_BITS;
+        int alloc = 1 + (FLINT_BITS != GMP_LIMB_BITS) + exp / GMP_LIMB_BITS
```

Comment:
Again we should not need the FLINT_BITS != GMP_LIMB_BITS case here. We don't support it if they are not equal.

Reply:
:-)